hawkbit-repository/hawkbit-repository-core/src/main/java/org/eclipse/hawkbit/repository/jpa/rsql/VirtualPropertyResolver.java

```diff
@@ -0,0 +1,74 @@
+/**
+ * Copyright (c) 2015 Bosch Software Innovations GmbH and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.repository.jpa.rsql;
+
+import java.time.Instant;
+
+import org.apache.commons.lang3.text.StrLookup;
+import org.apache.commons.lang3.text.StrSubstitutor;
+import org.eclipse.hawkbit.repository.jpa.TimestampCalculator;
+import org.eclipse.hawkbit.repository.rsql.VirtualPropertyReplacer;
+
+/**
+ * Adds macro capabilities to RSQL expressions that are used to filter for
+ * targets.
+ * <p>
+ * Some (virtual) properties do not have a representation in the database (in
+ * general these properties are time-related, or more explicitly, they deal with
+ * time intervals).<br>
+ * Such a virtual property needs to be calculated on Java-side before it may be
+ * used in a target filter query that is passed to the database. Therefore a
+ * placeholder is used in the RSQL expression that is expanded when the RSQL is
+ * parsed
+ * <p>
+ * A virtual property may either be a system value like the current date (aka
+ * <em>now_ts</em>) or a value derived from (tenant-specific) system
+ * configuration (e.g. <em>overdue_ts</em>).
+ * <p>
+ * Known values are:<br>
+ * <ul>
+ * <li><em>now_ts</em>: maps to system UTC time in milliseconds since Unix epoch
+ * as long value</li>
+ * <li><em>overdue_ts</em>: is a calculated value: <em>overdue_ts = now_ts -
+ * pollingInterval - pollingOverdueInterval</em>; pollingInterval and
+ * pollingOverdueInterval are retrieved from tenant-specific system
+ * configuration.</li>
+ * </ul>
+ */
+public class VirtualPropertyResolver extends StrLookup<String> implements VirtualPropertyReplacer {
+
+    private final TimestampCalculator timestampCalculator = new TimestampCalculator();
```

Comment:
I think it would be better if the `TimeStampCalculator` would be passed as constructor argument. Then you can get rid of the `getTimestampCalculator()` and easy for testing the `VirtualPropertyResolver` class.